[OpenTelemetry] Model changes to allow OpenTelemetry context propagation

stankiewicz opened the pull request:
Changes:
- dependencies definitions to otel
- WindowedValue interface change to expose context for propagation - context is not read by user
- changes are behind external_metadata experiment 

Next PRs:
- propagation to windmill sink
- reading from windmill reader
-  enabling default otel implementation for dataflow runner (v1 with propagation, v2 no propagation yet)

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] Mention the appropriate issue in your description (for example: `addresses #123`), if applicable. This will automatically add a link to the pull request in the issue. If you would like the issue to automatically close on merging the pull request, comment `fixes #<ISSUE NUMBER>` instead.
 - [ ] Update `CHANGES.md` with noteworthy changes.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://github.com/apache/beam/blob/master/CONTRIBUTING.md#make-the-reviewers-job-easier).

To check the build health, please visit [https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md](https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md)

GitHub Actions Tests Status (on master branch)
------------------------------------------------------------------------------------------------
[![Build python source distribution and wheels](https://github.com/apache/beam/actions/workflows/build_wheels.yml/badge.svg?event=schedule&&?branch=master)](https://github.com/apache/beam/actions?query=workflow%3A%22Build+python+source+distribution+and+wheels%22+branch%3Amaster+event%3Aschedule)
[![Python tests](https://github.com/apache/beam/actions/workflows/python_tests.yml/badge.svg?event=schedule&&?branch=master)](https://github.com/apache/beam/actions?query=workflow%3A%22Python+Tests%22+branch%3Amaster+event%3Aschedule)
[![Java tests](https://github.com/apache/beam/actions/workflows/java_tests.yml/badge.svg?event=schedule&&?branch=master)](https://github.com/apache/beam/actions?query=workflow%3A%22Java+Tests%22+branch%3Amaster+event%3Aschedule)
[![Go tests](https://github.com/apache/beam/actions/workflows/go_tests.yml/badge.svg?event=schedule&&?branch=master)](https://github.com/apache/beam/actions?query=workflow%3A%22Go+tests%22+branch%3Amaster+event%3Aschedule)

See [CI.md](https://github.com/apache/beam/blob/master/CI.md) for more information about GitHub Actions CI or the [workflows README](https://github.com/apache/beam/blob/master/.github/workflows/README.md) to see a list of phrases to trigger workflows.

stankiewicz committed:
bump opentelemetry version.

sjvanrossum commented on sdks/java/core/src/main/java/org/apache/beam/sdk/values/WindowedValues.java:
```diff
@@ -904,11 +987,14 @@ public <NewT> WindowedValueCoder<NewT> withValueCoder(Coder<NewT> valueCoder) {
     public void encode(WindowedValue<T> windowedElem, OutputStream outStream)
         throws CoderException, IOException {
-      encode(windowedElem, outStream, Context.NESTED);
+      encode(windowedElem, outStream, org.apache.beam.sdk.coders.Coder.Context.NESTED);
```
`Coder.Context` would suffice here I think?
```suggestion
      encode(windowedElem, outStream, Coder.Context.NESTED);
```

sjvanrossum commented on sdks/java/core/src/main/java/org/apache/beam/sdk/values/ValueInSingleWindow.java:
```diff
@@ -105,11 +109,14 @@ public static <T> Coder<T> of(
+    public void encode(
+        ValueInSingleWindow<T> windowedElem,
+        OutputStream outStream,
+        org.apache.beam.sdk.coders.Coder.Context context)
```
```suggestion
        Coder.Context context)
```

sjvanrossum commented on sdks/java/core/src/main/java/org/apache/beam/sdk/values/ValueInSingleWindow.java:
```diff
@@ -129,22 +140,27 @@ public void encode(ValueInSingleWindow<T> windowedElem, OutputStream outStream,
       Instant timestamp = InstantCoder.of().decode(inStream);
       BoundedWindow window = windowCoder.decode(inStream);
       PaneInfo paneInfo = PaneInfo.PaneInfoCoder.INSTANCE.decode(inStream);
+      io.opentelemetry.context.Context openTelemetryContext = null;
```
```suggestion
      @Nullable Context openTelemetryContext = null;
```

sjvanrossum commented on sdks/java/core/src/main/java/org/apache/beam/sdk/values/WindowedValues.java:
```diff
@@ -933,16 +1023,18 @@ public void encode(WindowedValue<T> windowedElem, OutputStream outStream, Contex
     @Override
     public WindowedValue<T> decode(InputStream inStream) throws CoderException, IOException {
-      return decode(inStream, Context.NESTED);
+      return decode(inStream, org.apache.beam.sdk.coders.Coder.Context.NESTED);
```
```suggestion
      return decode(inStream, Coder.Context.NESTED);
```

stankiewicz replied:
done

sjvanrossum commented on sdks/java/core/src/main/java/org/apache/beam/sdk/values/WindowedValue.java:
```diff
@@ -49,6 +50,9 @@ public interface WindowedValue<T> {
   String getRecordId();
 
+  @Nullable
+  Context getContext();
```
```suggestion
  Context getOpenTelemetryContext();
```
@kennknowles same here, right?

stankiewicz commented on buildSrc/src/main/groovy/org/apache/beam/gradle/BeamModulePlugin.groovy:
```diff
@@ -857,8 +857,13 @@ class BeamModulePlugin implements Plugin<Project> {
+        opentelemetry_api                           : "io.opentelemetry:opentelemetry-api:$opentelemetry_version",
         opentelemetry_bom                           : "io.opentelemetry:opentelemetry-bom-alpha:$opentelemetry_version-alpha", // alpha required by extensions
+        opentelemetry_context                       : "io.opentelemetry:opentelemetry-context:$opentelemetry_version",
+        opentelemetry_gcp_auth                      : "io.opentelemetry.contrib:opentelemetry-gcp-auth-extension:$opentelemetry_version-alpha",
```
thanks, fixed versioning.

sjvanrossum reviewed (changes requested):
Minor feedback, LGTM overall.

stankiewicz committed:
Model changes to allow OpenTelemetry context propagation
spotless.
rename fields, cleanup groovy
spotless
rename method, fix constructor invocation.

stankiewicz commented:
Run Java_GCP_IO_Direct PreCommit

stankiewicz committed:
add libraries

stankiewicz commented:
rebased & merged conflicts.

damccorm commented:
@kennknowles could you please take another look?

kennknowles reviewed (approved):
still LGTM

stankiewicz commented:
https://github.com/apache/beam/pull/38372 for follow up error prone fix.

Abacn commented on buildSrc/src/main/groovy/org/apache/beam/gradle/BeamModulePlugin.groovy:
```diff
@@ -857,6 +858,11 @@ class BeamModulePlugin implements Plugin<Project> {
         netty_transport_native_epoll                : "io.netty:netty-transport-native-epoll:$netty_version",
         opentelemetry_api                           : "io.opentelemetry:opentelemetry-api", // google_cloud_platform_libraries_bom sets version
         opentelemetry_bom                           : "io.opentelemetry:opentelemetry-bom-alpha:$opentelemetry_version-alpha", // alpha required by extensions
+        opentelemetry_context                       : "io.opentelemetry:opentelemetry-context", // google_cloud_platform_libraries_bom sets version
```
google_cloud_platform_libraries_bom unfortunately didn't set version for this dep, causing invalid pom
